src/statement.js

```diff
@@ -424,17 +442,23 @@ pp.parseVarId = function(decl) {
 // Parse a function declaration or literal (depending on the
 // `isStatement` parameter).
 
-pp.parseFunction = function(node, isStatement, allowExpressionBody) {
+pp.parseFunction = function(node, isStatement, allowExpressionBody, isAsync) {
   this.initFunction(node)
-  if (this.options.ecmaVersion >= 6)
+  if (this.options.ecmaVersion >= 6 && !isAsync)
     node.generator = this.eat(tt.star)
-  var oldInGen = this.inGenerator
+  if (this.options.ecmaVersion >= 8)
+    node.async = !!isAsync
+  if (isStatement && this.type === tt.name)
```

Comment:
Thank you for the good catch. I will fix this.

This was because of:

```js
// 'await' can be used as function name for function declarations.
async function await() {}

// but not for function expressions. (syntax error)
(async function await() {})

// Also, the function declaration at `export default` can be omitted their name.
export default function() {}
export default async function() {}